agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java

```diff
@@ -16,53 +16,80 @@
 
 package io.opentelemetry.auto.bootstrap;
 
+import io.opentelemetry.auto.bootstrap.instrumentation.api.Pair;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.ref.WeakReference;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.net.URLStreamHandler;
-import java.nio.file.NoSuchFileException;
 import java.security.Permission;
 import java.util.Enumeration;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.HashSet;
+import java.util.Set;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import lombok.extern.slf4j.Slf4j;
 
 @Slf4j
 public class InternalJarURLHandler extends URLStreamHandler {
-  private final Map<String, JarEntry> filenameToEntry = new HashMap<>();
-  private JarFile bootstrapJarFile;
+
+  private static final WeakReference<Pair<String, JarEntry>> NULL = new WeakReference<>(null);
+
+  private final String name;
+  private final FileNotInInternalJar notFound;
+  private final Set<String> packages = new HashSet<>();
```

Comment:
Realize it's the same as the previous pattern so optional update, we should probably set this to an `unmodifaibleSet` in the constructor instead of passing a mutable object around?

Reply:
This is only used within a very narrow context where (in the Datadog codebase, at least) we will not mutate the set once it's built, making the unmodifiable wrapper a pointless cost. If we wanted to present a view of this set to untrusted callers, we would have wrapped it.

Reply:
Note that in this fork, you're not actually using the set at all, so I recommend just removing it and saving a bit of space.

Reply:
thx! removed